modules/system/classes/VersionManager.php

```diff
@@ -197,6 +197,11 @@ public function removePlugin($plugin, $stopOnVersion = null, $stopCurrentVersion
                 $this->removeDatabaseScript($code, $history->version, $history->detail);
             }
 
+            $lastHistory = $this->getLastHistory($code);
+            if ($lastHistory) {
+                $this->setDatabaseVersion($code, $lastHistory->version);
+            }
+
```

Comment:
@prhost This particular code block and the new method you have added (`getLastHistory`) will result in a database query for each version rolled back in a plugin. I would instead, perhaps, have a `try..catch` block in this method surrounding the rollback functionality, with the `catch` block setting the database version number and then passing on the exception.

Reply:
I didn't understand it very well, could you explain it better? This `removePlugin` method of the `VersionManager` class is used by the `UpdateManager->rollbackPlugin()` which is used in several places in Octobercms, including also plugins like the Builder. I avoided making any exception shots because I don't know if those other places are waiting for that.

Reply:
@prhost Apologies - I'll try and explain it better:

That code above is doing 2 things right now:
- `getLastHistory` is querying the `system_plugin_history` table for the most recent update for a plugin after rolling back the current update. This is running a single read query.
- `setDatabaseVersion` is then running a single write query to update the plugin version for the plugin being rolled back in the `system_plugin_versions` table.

So effectively, every single version rolled back as part of the `removePlugin` method is running 2 queries on the database. If you say rollback 30 versions of a particular plugin, that is 60 queries.

Instead of doing this, I suggest you rely on the `foreach` loop which is already looping through the history, and just store the last history version as a variable until the end of the loop, and just use that to set the database version. With my suggestion of handling exceptions, you can simply wrap the entire loop in a try/catch block, and if an exception is to occur, you can break out of the loop and use the last history version variable to at least indicate it rolled back to that point.

I hope that makes sense :)

Reply:
Now I understand hehe. It really made a difference when it came to enough versions. I didn't even think about in commandline optimization because it is an environment for developers that uses few resources and that doesn't take too long, but it was better anyway.

Reply:
@prhost If we can optimise something, we will. We always want October to be the most finely-tuned piece of software out there ;)